src/python.c

```diff
@@ -1642,6 +1685,51 @@ canSort(MulticornPlanState * state, List *deparsed)
 	return result;
 }
 
+/*
+ * Call the can_pushdown_upperrel method from the python implementation, to
+ * determine whether upper relations can be pushed down to the corresponding
+ * data source to begin with.
+ *
+ * If yes, then also initialize some fields in MulticornPlanState needed for
+ * more granular conditional logic for assesing whether the particular query
+ * is suitable for pushdown.
+ */
+bool
+canPushdownUpperrel(MulticornPlanState * state)
+{
+    PyObject    *fdw_instance = state->fdw_instance,
+                *p_upperrel_pushdown,
+                *p_object;
+    bool pushdown_upperrel = false;
+
+    p_upperrel_pushdown = PyObject_CallMethod(fdw_instance, "can_pushdown_upperrel", "()");
+    errorCheck();
+
+    if (p_upperrel_pushdown != NULL && p_upperrel_pushdown != Py_None)
+    {
+        /* Determine whether the FDW instance supports GROUP BYs */
+        p_object = PyMapping_GetItemString(p_upperrel_pushdown, "groupby_supported");
+        if (p_object != NULL && p_object != Py_None)
+		{
+            state->groupby_supported = PyObject_IsTrue(p_object);
+            Py_DECREF(p_object);
+		}
+
+        /* Determine which aggregation functions are supported */
+        p_object = PyMapping_GetItemString(p_upperrel_pushdown, "agg_functions");
+        if (p_object != NULL && p_object != Py_None)
+		{
+            state->agg_functions = PyMapping_Keys(p_object);
```

Comment:
I don't think you ever DECREF `state->agg_functions`, so this will slowly leak. I'd extract the contents into a separate List here and get rid of the PyObject here so that you also don't have to mess with the Python API in `foreign_expr_walker`.

Reply:
Good catch. I initially tried the route you mentioned but was stuck extracting Python Unicode objects into a PG List, so I went with this instead. Let me get back at this.

Reply:
Ok, I've now added storing of supported agg functions to a List.